### PR TITLE
[MRESOLVER-445] Simplify session handling, move out logic from session builder

### PR DESCRIPTION
**Pull request by cstamas**
Simplify session handling (copy is gone), also move out logic from session builder to make it reusable.

---

https://issues.apache.org/jira/browse/MRESOLVER-445

### CODE REVIEW

**Comment by gnodet** [comment]
@cstamas have you looked at the possibility to add mutators that would return a new `RepositorySession` instance instead of modifying the existing one in-place.  This would allow preserving the immutability, as this feels it goes a bit backward...

If that does not work, it should be possible to implement a wrapper that would simply delegate to a mutable field.
```
public class DefaultMutableSession implements MutableRepositorySession {
  private RepositorySession session;
  public DefaultMutableSession(RepositorySession session) {
    this.session = session;
  }
  public void setCache(RepositoryCache cache) {
    this.session = session.withCache(cache);
  }
  ...
}
```

**Comment by cstamas** [comment]
Actually scratch all that above, I changed my mind. Soon will push almost totally opposite change to the commit that introduces MutableSession.

File: maven-resolver-api/src/main/java/org/eclipse/aether/RepositorySystemSession.java [hunk 1]
```diff
@@ -391,20 +378,20 @@ interface SessionBuilder {
...
         /**
-         * Sets the cache the repository system may use to save data for future reuse during the session.
+         * Sets the custom session data supplier associated with this session.
          *
-         * @param cache The repository cache, may be {@code null} if none.
+         * @param dataSupplier The session data supplier, may not be {@code null}.
          * @return This session for chaining, never {@code null}.
          */
-        SessionBuilder setCache(RepositoryCache cache);
+        SessionBuilder setSessionDataSupplier(Supplier<SessionData> dataSupplier);
```
[the remark this answers is not in the document]
Comment (cstamas) [reply maven-resolver-api/src/main/java/org/eclipse/aether/RepositorySystemSession.java:394]: Will add a warning about this above to setCache/setData methods, and will merge this PR unless some other concern crops up.

**Comment by jira-importer** [comment]
Resolve #1119

**Pull request merged**